File: src/decoders/notebook-delay.ts
import type { NotebookDelay } from "~/models/notebook-delay";
import { decodePronoteDate } from "./pronote-date";

export const decodeNotebookDelay = (delay: any): NotebookDelay => {
  const isReasonUnknown = delay.estMotifNonEncoreConnu;
  const justified = delay.justifie;

  return {
    id: delay.N,
    date: decodePronoteDate(delay.date.V),
    minutes: delay.duree,
    justified,
    justification: justified && delay.justification,
    shouldParentsJustify: delay.aJustifierParParents,
    administrativelyFixed: delay.reglee,
    isReasonUnknown,
    reason: !isReasonUnknown && delay.listeMotifs.V[0].L
  };
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/decoders/notebook-delay.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.models.NotebookDelay
import kotlinx.serialization.json.*

fun decodeNotebookDelay(delay: JsonObject): NotebookDelay {
  val isReasonUnknown = delay["estMotifNonEncoreConnu"]!!.jsonPrimitive.boolean
  val justified = delay["justifie"]!!.jsonPrimitive.boolean

  return NotebookDelay(
      id = delay["N"]!!.jsonPrimitive.content,
      date = decodePronoteDate(delay["date"]!!.jsonObject["V"]!!.jsonPrimitive.content),
      minutes = delay["duree"]!!.jsonPrimitive.int,
      justified = justified,
      justification = if (justified) delay["justification"]!!.jsonPrimitive.content else null,
      shouldParentsJustify = delay["aJustifierParParents"]!!.jsonPrimitive.boolean,
      administrativelyFixed = delay["reglee"]!!.jsonPrimitive.boolean,
      isReasonUnknown = isReasonUnknown,
      reason =
          if (!isReasonUnknown)
              delay["listeMotifs"]!!
                  .jsonObject["V"]!!
                  .jsonArray[0]!!
                  .jsonObject["L"]!!
                  .jsonPrimitive
                  .content
          else null)
}
